File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        
        int product = 1;
        int flag = 0;
        
        for(int i = 0; i < nums.size(); i++) {
            if (nums[i] == 0) {
                flag += 1;
                continue;
            }
            product *= nums[i];
        }
        
        vector<int> result(nums.size(), product);
        
        if (flag == 0) {
            for(int i = 0; i < nums.size(); i++) {
                result[i] /= nums[i];
            }
        }
        else if (flag > 1) {
            for(int i = 0; i < nums.size(); i++) {
                result[i] = 0;
            }
        }
        else {
            for(int i = 0; i < nums.size(); i++) {
                if (nums[i] == 0) {
                    result[i] = product;
                }
                else {
                    result[i] = 0;
                }
            }
        }
        
        return result;
        
        
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        if len(nums) <= 1:
            return nums

        product = 1
        zeroFlag = 0
        
        for num in nums:
            if num == 0:
                zeroFlag += 1
            else:
                product *= num
        
        if zeroFlag > 1:
            return [0] * len(nums)
        
        for idx, num in enumerate(nums):
            if zeroFlag == 1 and num != 0:    
                nums[idx] = 0
            elif zeroFlag == 1 and num == 0:
                nums[idx] = product
            else:
                nums[idx] = int(product/num)
        
        return nums